### PR TITLE
JWT authentication of Issuers & Receivers

### PR DESCRIPTION
**Pull request by Khukhuna**
Part of GCI 2018 :tada: :balloon: 

**[Video preview](https://www.youtube.com/watch?v=KK3P9I8xnrw&feature=youtu.be)** of sending post request to add issuers & receivers

**Changes:**
- Registration of Certificate Issuers & Receivers
- Model tests.

Right now server returns only username and name as jwt, please let me know if i should add anything else. :smile:

### CODE REVIEW

**Comment by Khukhuna** [comment]
Travis is failing because of hardcoding in tests.py file. :cry: 
This is very unimportant error in this case.
I will research solution, but If there's any way to exclude tests.py file from codacy, please let me know.

File: api/views.py [hunk 1]
```diff
@@ -9,9 +16,46 @@
 from api import (
     models,
     serializers,
-    permissions as custom_permissions,
+    permissions as custom_permissions
 )
 
+@authentication_classes([])
+@permission_classes([])
+class Registration(APIView):
```
[the remark this answers is not in the document]
Comment (Khukhuna) [reply api/views.py]: Yes :smile: I'll do refactor

**Comment by Khukhuna** [comment]
Did the code refactor. Now it uses generic CreateAPIView.

File: api/views.py [hunk 2]
```diff
@@ -9,9 +12,46 @@
...
+
+    @classmethod
+    def create(self, request, *args, **kwargs):
+        serializer = self.serializer_class(data=request.data)
+        try:
+            if serializer.is_valid(raise_exception=True):
+                serializer.create(request.data)
+                jwt_token = {'token': jwt.encode(serializer.data, config('SECRET_KEY'))}
+                return Response(jwt_token, status=status.HTTP_201_CREATED)
+        except:
```
Comment (aashutoshrathi) [inline api/views.py]: Specify error

File: api/urls.py [hunk 3]
```diff
@@ -3,6 +3,8 @@
 from api import views
 
 urlpatterns = [
+    path('user/', views.StudentCreation.as_view()),
```
Comment (aashutoshrathi) [inline api/urls.py]: Is it a good idea to keep the path as `user/`?

Comment (Khukhuna) [reply api/urls.py]: I'm open for suggestions :/ I'm so bad at naming variables, paths etc. :joy:

Comment (aashutoshrathi) [reply api/urls.py]: I mean we have `user/<int>` for UserDetailView. :sweat_smile:

Comment (Khukhuna) [reply api/urls.py]: maybe we can have like:  `http://domain.com/api/user/register` ? You tell me and I'll change it :joy:

**Comment by Khukhuna** [comment]
This codacy errors are useless in this case & can be ignored.
or we can exclude test files from codacy.

File: api/urls.py (hunk 3 above)
Comment (aashutoshrathi) [reply api/urls.py]: @jatinkatyal13 views?

Comment (jatinkatyal13) [reply api/urls.py]: https://domain.com/api/<org|user> [POST] should register the user and user/<int:pk> [GET] should fetch the Details and [POST] should patch it

**Comment by Khukhuna** [comment]
As requested, changed paths to:
 - **`/api/user`**           - [POST] - Registers user
 - **`/api/user/<int>`**  - Fetch the user details
 - **`/api/org/`**            - Registers organization

**Comment by junron** [comment]
I've been tinkering around with your code, trying to fix the failing Codacy tests.
I'm not sure why you used `User.objects.create`  and I couldn't find any docs on that,
but using `User.objects.create_user` seems to work, and it takes positional arguments, which removes the need for the keyword `password`.

Hence, this works, and eliminates the use of `password`
`user = User.objects.create_user("jbossAdmin",None,"jbossAdminsPassword")`

Of course, another possibility would be that the string `password` within the password argument value is not allowed, in which case it can be fixed easily.

Hope this helps!

**Comment by Khukhuna** [comment]
@junron It works. Thank you!

**Comment by jatinkatyal13** [comment]
> As requested, changed paths to:
> 
> * **`/api/user`**           - [POST] - Registers user
> * **`/api/org/`**            - [POST] - Registers organization
> * **`/api/user/<int>`**  -  [GET] - Fetch the user details

Amazing ! 🚀

**Pull request merged**